controllers/webhook/webhook_controller.go

```diff
@@ -0,0 +1,220 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webhook
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/istio-ecosystem/sail-operator/api/v1alpha1"
+	"github.com/istio-ecosystem/sail-operator/pkg/constants"
+	"github.com/istio-ecosystem/sail-operator/pkg/reconciler"
+	admissionv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	defaultPeriodSeconds  = 3 // matches the period in the istiod chart
+	defaultTimeoutSeconds = 5 // matches the timeout in the istiod chart
+)
+
+// overrides the default dial context; only used in unit tests
+var customDialContext func(ctx context.Context, network, addr string) (net.Conn, error)
+
+// Reconciler checks the readiness of MutatingWebhookConfiguration pointing to a remote Istio control plane
+type Reconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	probe  func(context.Context, *admissionv1.MutatingWebhookConfiguration) (bool, error)
+}
+
+func NewReconciler(client client.Client, scheme *runtime.Scheme) *Reconciler {
+	return &Reconciler{
+		Client: client,
+		Scheme: scheme,
+		probe:  doProbe,
+	}
+}
+
+// +kubebuilder:rbac:groups=admissionregistration.k8s.io,resources=mutatingwebhookconfigurations,verbs=get;list;watch;create;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
+func (r *Reconciler) Reconcile(ctx context.Context, webhook *admissionv1.MutatingWebhookConfiguration) (ctrl.Result, error) {
+	isReady, err := r.probe(ctx, webhook)
+	if err != nil {
+		isReady = false
+	}
+
+	if webhook.Annotations == nil {
+		webhook.Annotations = make(map[string]string)
+	}
+	webhook.Annotations[constants.WebhookReadinessProbeStatusAnnotationKey] = strconv.FormatBool(isReady)
```

Comment:
The webhook status is handled by the webhook controller, not by the istiorevision controller. This is because the status must be checked periodically and not when some owned resource is updated (this is the case in the istiorevision controller). The webhook controller shouldn't update the IstioRevision status.